File: src/core/source-detector.ts
import { existsSync, statSync } from 'node:fs'
import { extname, isAbsolute, resolve } from 'node:path'

export interface DetectedSource {
  type: 'file' | 'folder' | 'url' | 'github'
  input: string
}

export function detectSource(input: string): DetectedSource {
  if (input.startsWith('https://github.com/') || input.startsWith('http://github.com/')) {
    return { type: 'github', input }
  }
  if (input.startsWith('http://') || input.startsWith('https://')) {
    return { type: 'url', input }
  }
  const absPath = isAbsolute(input) ? input : resolve(process.cwd(), input)
  if (existsSync(absPath)) {
    const s = statSync(absPath)
    if (s.isDirectory()) return { type: 'folder', input: absPath }
    return { type: 'file', input: absPath }
  }
  return { type: 'url', input }
}

File: python/contextlane/core/source_detector.py
import os
from pathlib import Path


def detect_source(input_str: str) -> tuple:
    """Returns (type, resolved_input)."""
    if input_str.startswith("https://github.com/") or input_str.startswith("http://github.com/"):
        return ("github", input_str)

    if input_str.startswith("http://") or input_str.startswith("https://"):
        return ("url", input_str)

    p = Path(input_str)
    if p.exists():
        if p.is_dir():
            return ("folder", str(p.resolve()))
        return ("file", str(p.resolve()))

    return ("url", input_str)
